Prepare Kinto.elm to be released as a standalone package

magopian opened the pull request:
- [x] [Write helpful documentation with examples](http://package.elm-lang.org/help/design-guidelines#write-helpful-documentation-with-examples)
- [x] [Design for a concrete use case](http://package.elm-lang.org/help/design-guidelines#design-for-a-concrete-use-case)
- [x] [The data structure is always the last argument](http://package.elm-lang.org/help/design-guidelines#the-data-structure-is-always-the-last-argument)
- [x] [Keep tags and record constructors secret](http://package.elm-lang.org/help/design-guidelines#write-helpful-documentation-with-examples)
- [x] [Make sure the naming is good](http://package.elm-lang.org/help/design-guidelines#naming)

magopian committed:
The data structure is always the last argument

magopian committed:
First attempt at a brand new API

magopian committed:
small variation on the 'new api'

magopian committed:
Another attempt at a simpler version of the API

glasserc commented on src/Kinto.elm:
```diff
@@ -272,262 +282,307 @@ extractKintoError statusCode statusMsg body =
+        |> endpointUrl config.baseUrl
+        |> HttpBuilder.get
+        |> HttpBuilder.withHeaders config.headers
+        |> HttpBuilder.withExpect (Http.expectJson config.recordDecoder)
```
Here's an idea: maybe we can push this a little bit further. Let's have the requests include something to reflect the kind of request they're making (and thus, the kind of response they will get). So, how about having `get` produce an `HttpBuilder.RequestBuilder (GetResponse a)`? This could include permissions or whatever other non-record stuff is in a GET response. Then, you could add a `Kinto.justRecord : HttpBuilder.RequestBuilder (GetResponse a) -> HttpBuilder.RequestBuilder a`. I think this will also make it harder to accidentally tag a CreateRecord request with a DeleteRecord message type. What do you think? Too much?

magopian committed:
Yet another attempt

n1k0 commented on src/Kinto.elm:
```diff
@@ -553,13 +551,47 @@ send tagger builder =
+update : Body -> ResourceConfig -> HttpBuilder.RequestBuilder Decode.Value
+update body config =
+    endpointUrl config.baseUrl config.endpoint
+        |> HttpBuilder.patch
```
We'll have either to deal with more options or another function because updating may be achieved using POST too

n1k0 commented on src/Model.elm:
```diff
@@ -67,13 +67,18 @@ config =
+        (Kinto.Basic "test" "test")
+
+
+recordListConfig =
```
I feel like the basic server config should be distinguished from the resource configuration (eg. here `recordConfig` and `recordListConfig` share a lot). And to be entirely honest, I also feel like users shouldn't have to deal with resource endpoint configuration at all.

Ideally as a lib user I'd like a slighly higher level API allowing to simply request for resources using dedicated functions where I could pass a type mapper to process the response and possibly some options (eg. safe requests or a patch flag and so on)

n1k0 commented on src/Model.elm:
```diff
@@ -45,19 +45,19 @@ type Msg
+    | EditRecordResponse (Result Kinto.Error Record)
     | DeleteRecord RecordId
-    | DeleteRecordResponse (Result Kinto.Error Value)
+    | DeleteRecordResponse (Result Kinto.Error Record)
```
I really like this, providing a target type is so much more comfortable to use!

n1k0 commented on src/Model.elm:
```diff
@@ -219,18 +210,31 @@ subscriptions model =
+    recordListConfig
+        |> Kinto.get
+        |> Kinto.withDecoder (decodeData decodeRecords)
+        |> Kinto.send FetchRecordsResponse
```
I like this pipeline style a lot.

n1k0 commented on src/Model.elm:
```diff
@@ -219,18 +210,31 @@ subscriptions model =
+
+decodeData : Decoder a -> Decoder a
+decodeData decoder =
+    field "data" decoder
```
It's odd to me that this has to be defined in userland; I really feel like this should be handled by the lib itself, as a core, hidden implementation detail.

n1k0 commented on src/Model.elm:
```diff
@@ -219,18 +210,31 @@ subscriptions model =
 
 fetchRecord : RecordId -> Cmd Msg
```
Why are these living in Model.elm and not in Kinto.elm? I may be missing something obvious, or you're possibly in the middle of migrating them there, but to me these should be fully part of the core API of a Kinto client library :)

n1k0 reviewed (commented):
Here are a few random thoughts while reading the code, please take it with a grain of salt as I've been far from the ongoing implementation effort for a while!

magopian committed:
Are we close?

n1k0 commented on src/Model.elm:
```diff
@@ -321,3 +231,72 @@ updateRecord formData record =
+        |> Kinto.send FetchRecordResponse
+
+
+fetchRecords : Cmd Msg
```
Nit: `fetchRecordList` maybe, as `fetchRecords` is visually harder to distinguish from `fetchRecord` at a quick glance

magopian committed:
Add filtering

magopian committed:
Add sorting

magopian committed:
Don't use 'Kinto' in the names if they're not needed

magopian committed:
Encode data automatically, add the 'replace' API

magopian committed:
Improve the README to provide info for the future elm package, and an example

magopian committed:
Add a CHANGELOD and LICENSE file

magopian committed:
Remvoe unused imports

magopian committed:
Add documentation

magopian commented:
To preview the documentation, go to http://package.elm-lang.org/help/docs-preview and upload the content from http://0bin.net/paste/-7JPV5kotmnRiZqQ#Gm5R0eNQymXLy1hlo84BwYHvly6KSj-ykFqnZ4iOPN2 into a documentation.json file.

magopian committed:
Modify the file structure and move all the things to make it a publishable elm package

magopian committed:
elm packages start at version 1.0.0

n1k0 commented on README.md:
```diff
@@ -1,41 +1,117 @@
+
+decodeTodo : Decode.Decoder Todo
+decodeTodo =
+    (Decode.map4 Todo
```
Nit: do we need to wrap the whole statement within parenthesis?

magopian replied:
well spotted, thanks

n1k0 commented on README.md:
```diff
@@ -1,41 +1,117 @@
+-- Get all the Todos
+
+
+getTodos : Cmd Msg
```
Nit: `getTodoList`?

n1k0 commented on examples/Example.elm:
```diff
@@ -0,0 +1,149 @@
+
+
+addTodo : Maybe String -> Maybe String -> Cmd Msg
+addTodo title description =
```
I'd naively expect `addTodo` to accept a `Todo` record, but I understand that's not compatible with the expectation it has an `id` and a `last_modified`. We could have a `TodoFields` type to distinguish between a Kinto record and the custom data it contains, eg.

```elm
type TodoFields = {
  title: String,
  description: String
}

type alias Todo = {
  id: String,
  last_modified: Int,
  fields: TodoFields,
}
```

That way, `addRecord` could just accept a `TodoFields`. Note that encodeData could then accept the same. I'd find it more clear, but also slightly overkill, so... your call.

glasserc replied:
The [Yesod framework](http://www.yesodweb.com/) solves this problem by defining an `Entity a` type that is essentially `(Id, a)`. So here you would have a `Todo` record, but a `Todo` that was in the DB would be an `Entity Todo`. They write about it [here](http://www.yesodweb.com/book/persistent#persistent_insert).

n1k0 replied:
That sounds rather smart, I like it.

magopian replied:
I agree with both those comments, but I'd rather not implement them here: It's an example file that is used as the "base" for the example in the README, and I'd like it to be as minimal as possible.

On that note, I'm not entirely happy with the way it is at the moment: I have no way to make sure that the Example.elm file compiles properly (i'll try to add an entry in travis or something), and even if it does, I need to remember to mirror all the changes in it that are also in the README in both the files...

n1k0 commented on examples/Example.elm:
```diff
@@ -0,0 +1,149 @@
+            { model | error = Nothing, todos = todoList } ! []
+
+        TodosFetched (Err error) ->
+            { model | error = Just <| toString error } ! []
```
I wonder if there's a way to do this, and if that would be desirable to showcase in example code:

```elm
update message model =
    case message of
        -- ...
        _ (Err error) ->
            { model | error = Just <| toString error } ! []
```

glasserc replied:
It sure does feel like there's a lot of boilerplate here. Since we don't actually do anything different on the different error cases, we could make `TodoAdded` and `TodosFetched` only handle the `Ok` case, and define a single `CommunicationError` type to handle the shared `Error` cases.

magopian replied:
Your proposition doesn't work sadly @n1k0 i just tried :/
@glasserc as explained above, I'd rather have a very simple, concise and straightforward (should I say "elm-ic"?) code as it's for an example, so I think I'll pass on this one. Definitely need to give it a try though in the future.

n1k0 commented on package.json:
```diff
@@ -1,31 +1,25 @@
-    "start": "node_modules/.bin/elm-reactor",
-    "debug": "node_modules/.bin/elm-live src/Main.elm --output=html/live/app.js --dir=html/live -- --debug",
-    "live": "node_modules/.bin/elm-live src/Main.elm --output=html/live/app.js --dir=html/live",
-    "publish-to-gh-pages": "npm run build && cp html/index.html build/ && node_modules/.bin/gh-pages --dist build/",
```
Don't we want to keep the ability to publish the example app there?

magopian replied:
so, it turns out that if you do `npm run build` for example in the root folder, it'll use the `elm-package.json` which is at the same level, instead of using the one in `examples/`, and thus fails.

That's why I created a project `package.json` file in the `examples/` folder, which has all the cool stuff, and a limited/simpler "library" one in the root folder.

n1k0 reviewed (approved):
Apart a few minor remarks this looks awesome! :+1:

glasserc commented on src/Kinto.elm:
```diff
@@ -64,22 +65,69 @@ type Endpoint
+    { itemEndpoint : String -> Endpoint
+    , listEndpoint : Endpoint
+    , itemDecoder : Decode.Decoder a
+    , listDecoder : Decode.Decoder (List a)
```
I really like the way you've separated the verbs get/create/update/delete from the things being manipulated. Some random thoughts:

- Do we really need `listDecoder`? It seems like it should always just be equivalent to use `Decode.list resource.itemDecoder`.
- It's conceptually possible to have a `Resource (BucketEndpoint) (CollectionListEndpoint someBucket)`, although that doesn't make any sense. It's not actually possible for a user of this module to construct something like that, but it feels weird to me. I think fixing this would be a little annoying -- you'd have to separate singular endpoints from plural endpoints, and define URL mappings for both, so it might not be worth it.
- Going the other way, does the `Endpoint` type still have value? We no longer expose it, nor even need to expose it. I guess I still think it provides some conceptual clarity over just using strings..
- I feel like there's a little bit of ... redundancy, almost, between specifying a bucket, a collection, or a record. There's no way to link records to the collections that contain them, nor those collections to their buckets. But I can't think of any way to solve that which also provides any kind of value, so maybe that's fine.

magopian replied:
Yeah, appart from being "more explicit", the `listDecoder` is pretty useless.
I agree with your concerns, especially regarding the `Endpoint`s: there's another concern I had for a long time, and don't know yet how to deal with... you can call `getList` with an item endpoint, it'll fail at runtime while trying to decode a list instead of a single item...

I guess there's still a lot more work left on this API, maybe for a further iteration?

glasserc reviewed (commented):
I think this is a great step forward!

magopian committed:
Address review comments

magopian committed:
Add some more things to test for travis

magopian commented:
Ok, addressed (some of the) comments, added extra stuff to test for Travis, I thing we're good to merge. Thanks for all the feedback @glasserc and @n1k0!
